### PR TITLE
Copy relative or full path(s) to clipboard

### LINKED ISSUE #10158: Copy path from repository root
### Feature description

Please add "Copy relative path" in the same way as it shown in Diff files list.
image

As example JetBrains Rider has the same feature: File Menu -> Copy Path -> Path From Repository Root
and it's very useful

### Environment

- Git Extensions 3.5.4.12724
- Build 65f01f39982ccb121c4df7d4b00b506939553958
- Git 2.37.0.windows.1
- Microsoft Windows NT 10.0.19044.0
- .NET Framework 4.8.9035.0
- DPI 120dpi (125% scaling)

### PR DESCRIPTION
**Pull request by mstv**
Fixes #10158

## Proposed changes

- Copy file paths in variants (relative / full, native / POSIX / WSL / Cygwin) in RevisionDiffControl, RevisionFileTreeControl (and invisible in FormDiff)
- `RevisionDiffControl.Designer`: Reorder using Designer

**Note:** The parent menu item "Copy path(s)" can be clicked. Then the (bold) default action "Copy full path(s) - native" is executed.

### Before

image

### After

image

### Before

image

### After

image

## Test methodology 

- manual

## Test environment(s) 

- Git Extensions 33.33.33
- Build 0c47ce4a91b55b6e8603cece1d5e7eaf4f2b34d2 (Dirty)
- Git 2.39.2.windows.1
- Microsoft Windows NT 10.0.19045.0
- .NET 6.0.12
- DPI 96dpi (no scaling)
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

## Merge strategy

Please do **not squash** merge this PR!

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 1]
```diff
@@ -1701,35 +1701,6 @@ private void editGitAttributesToolStripMenuItem_Click(object sender, EventArgs e
...
-                {
-                    fileNames.AppendLine();
-                }
-
-                fileNames.Append(path.ToNativePath());
-            }
-
-            ClipboardUtil.TrySetText(fileNames.ToString());
-        }
-
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormBrowse.cs:1732]: ❤️

File: GitUI/CommandsDialogs/Menus/ToolStripMenuItemEx.cs [hunk 2]
```diff
@@ -14,6 +14,9 @@ internal abstract class ToolStripMenuItemEx : ToolStripMenuItem, ITranslate
         protected GitUICommands UICommands
             => (_getUICommands ?? throw new InvalidOperationException("The button is not initialized")).Invoke();
 
+        protected GitModule Module
```
Comment (RussKie) [inline GitUI/CommandsDialogs/Menus/ToolStripMenuItemEx.cs:17]: Please add xml-docs

File: GitUI/CommandsDialogs/Menus/CopyPathsToolStripMenuItem.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            }
+        }
+
+        private static string GetFilePaths(IEnumerable<string?> selectedFilePaths, string prefixDir, Func<string, string> convertPath)
+        {
+            return selectedFilePaths
+                .Where(path => !string.IsNullOrEmpty(path))
+                .Select(path => convertPath(Path.Combine(prefixDir, path)))
+                .Join(Environment.NewLine);
+        }
```
Comment (RussKie) [inline GitUI/CommandsDialogs/Menus/CopyPathsToolStripMenuItem.cs:67]: [nit] Please add instance and static methods before event handlers.

I generally use the original .NET Framework guidelines as a basis, https://docs.microsoft.com/en-us/archive/blogs/brada/internal-coding-guidelines#28-file-organization.

File: GitCommands/PathUtil.cs [hunk 4]
```diff
@@ -29,6 +29,39 @@ public static class PathUtil
...
+            path = path.ToPosixPath();
+            Match match = Regex.Match(path, "^([A-Za-z]):(.*)$");
+            if (match.Groups.Count != 3)
+            {
+                return path;
+            }
+
+            return $"{prefix}{match.Groups[1].Value.ToLowerInvariant()}{match.Groups[2]}";
+        }
+
```
Comment (RussKie) [inline GitCommands/PathUtil.cs:64]: It would be great to add tests for these.

File: GitUI/CommandsDialogs/Menus/ToolStripMenuItemEx.cs [hunk 5]
```diff
@@ -14,6 +14,9 @@ internal abstract class ToolStripMenuItemEx : ToolStripMenuItem, ITranslate
...
+            => UICommands.Module;
```
Comment (RussKie) [inline GitUI/CommandsDialogs/Menus/ToolStripMenuItemEx.cs:18]: Is this used here, or somewhere else?
image

File: GitUI/CommandsDialogs/Menus/CopyPathsToolStripMenuItem.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        public CopyPathsToolStripMenuItem Initialize(Func<GitUICommands> getUICommands, Func<IEnumerable<string?>> getSelectedFilePaths)
+        {
+            Initialize(getUICommands);
+            _getSelectedFilePaths = getSelectedFilePaths;
+            return this;
+        }
+
+        private void CopyFullPathsNativeToolStripMenuItem_Click(object sender, EventArgs e)
+        {
+            ((ToolStripItem)sender).GetCurrentParent().Hide();
```
Comment (RussKie) [inline GitUI/CommandsDialogs/Menus/CopyPathsToolStripMenuItem.cs:26]: Is there a reason we can't use `this.GetCurrentParent().Hide();`?

**Review by RussKie — CHANGES_REQUESTED** [review]
👍

File: GitUI/CommandsDialogs/Menus/ToolStripMenuItemEx.cs (hunk 2 above)
Comment (mstv) [reply GitUI/CommandsDialogs/Menus/ToolStripMenuItemEx.cs:20]: Done

File: GitUI/CommandsDialogs/Menus/CopyPathsToolStripMenuItem.cs (hunk 3 above)
Comment (mstv) [reply GitUI/CommandsDialogs/Menus/CopyPathsToolStripMenuItem.cs]: moved

File: GitCommands/PathUtil.cs (hunk 4 above)
Comment (mstv) [reply GitCommands/PathUtil.cs:67]: caught up on

File: GitUI/CommandsDialogs/Menus/ToolStripMenuItemEx.cs (hunk 5 above)
Comment (mstv) [reply GitUI/CommandsDialogs/Menus/ToolStripMenuItemEx.cs:21]: Yes, here. I expect more usages when we factor out more submenus. We use to have a `Module` property everywhere.

File: GitUI/CommandsDialogs/Menus/CopyPathsToolStripMenuItem.cs (hunk 6 above)
Comment (mstv) [reply GitUI/CommandsDialogs/Menus/CopyPathsToolStripMenuItem.cs]: No, there is none :)

File: GitCommands/PathUtil.cs [hunk 7]
```diff
@@ -29,6 +29,39 @@ public static class PathUtil
...
+        [return: NotNullIfNotNull("path")]
+        public static string? ToMountPath(this string? path, string prefix)
+        {
+            if (path is null)
+            {
+                return null;
+            }
+
+            path = path.ToPosixPath();
+            Match match = Regex.Match(path, "^([A-Za-z]):(.*)$");
```
[the remark this answers is not in the document]
Comment (mstv) [reply GitCommands/PathUtil.cs]: done

**Review by gerhardol — APPROVED** [review]
Have not run, but all looks fine

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
👍

**Pull request merged**